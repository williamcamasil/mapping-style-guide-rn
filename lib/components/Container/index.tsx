import React from 'react';
import { View, ViewProps, ViewStyle } from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';

export type ContainerPropsType = ViewProps & {
  theme: AppThemeType;
  padding?: ViewStyle['padding'];
  paddingVertical?: ViewStyle['padding'];
  paddingHorizontal?: ViewStyle['padding'];
  paddingLeft?: ViewStyle['padding'];
  paddingRight?: ViewStyle['padding'];
  paddingTop?: ViewStyle['padding'];
  paddingBottom?: ViewStyle['padding'];
  fillParent?: boolean;
  horizontal?: boolean;
  centered?: boolean;
  backgroundColor?: ViewStyle['backgroundColor'];
  blockActions?: boolean;
  justifyContent?: ViewStyle['justifyContent'];
  alignItems?: ViewStyle['alignItems'];
};

const Container: React.FC<ContainerPropsType> = ({
  children,
  style,
  theme,
  padding,
  paddingVertical,
  paddingHorizontal,
  paddingLeft,
  paddingRight,
  paddingTop,
  paddingBottom,
  fillParent,
  horizontal,
  centered,
  backgroundColor,
  blockActions,
  justifyContent,
  alignItems,
  ...others
}) => {
  const containerStyles = useViewStyles(() => [
    {
      flex: fillParent ? 1 : undefined,
      flexDirection: horizontal ? 'row' : undefined,
      justifyContent: justifyContent ?? (centered && horizontal ? 'center' : undefined),
      alignItems: alignItems ?? (centered && !horizontal ? 'center' : undefined),
      backgroundColor: backgroundColor || theme.colors.neutralWhite,
      padding,
      paddingVertical,
      paddingHorizontal,
      paddingLeft,
      paddingRight,
      paddingTop,
      paddingBottom,
    },
    style,
  ], [
    backgroundColor,
    centered,
    fillParent,
    horizontal,
    padding,
    paddingBottom,
    paddingHorizontal,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingVertical,
    style,
    theme.colors.neutralWhite,
    justifyContent,
    alignItems,
  ]);

  return (
    <View pointerEvents={blockActions ? 'none' : 'auto'} style={containerStyles} {...others}>
      {children}
    </View>
  );
};

Container.defaultProps = {
  fillParent: true,
};

export default withTheme(Container);
