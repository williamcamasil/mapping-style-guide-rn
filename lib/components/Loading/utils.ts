import { ColorValue } from 'react-native';

import Color from 'color';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export function getCircleBackgroundColor(
  theme: AppThemeType,
  color: ColorPaletesType,
): ColorValue | undefined {
  return theme.colors[color];
}

export function getCircleBorderColor(
  theme: AppThemeType,
  color: ColorPaletesType,
): ColorValue | undefined {
  const colorValue = theme.colors[color];

  return Color(colorValue).alpha(0.6).hsl().toString();
}
