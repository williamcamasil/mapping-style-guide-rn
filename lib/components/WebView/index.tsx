import React, { ComponentType, useCallback, useRef } from 'react';
import { View, StyleSheet } from 'react-native';

import { useViewStyles } from '../../hooks';

/*
 * Importa a lib de forma inline e opcional para não obrigar os demais repositórios
 * a instalarem o também.
 */
// @ts-ignore
type WebViewProps = import('react-native-webview').WebViewProps;
// @ts-ignore
type WebViewProgressEvent = import('react-native-webview/lib/WebViewTypes').WebViewProgressEvent;
let RNWebView: ComponentType<WebViewProps>;
/* istanbul ignore next */
try {
  if (require('react-native-webview')) {
    RNWebView = require('react-native-webview').default;
  }
} catch (err) {
  if (process.env.NODE_ENV !== 'test') {
    console.warn(err);
  }
}

export type WebViewPageChangeEventType = {
  title: string;
  url: string;
};

export type WebViewPropsType = WebViewProps & {
  renderLoading?: () => React.ReactNode;
  onPageChange?: (event: WebViewPageChangeEventType) => void;
};

const styles = StyleSheet.create({
  webView: {
    flex: 1,
  },
});

const WebView: React.FC<WebViewPropsType> = ({
  renderLoading, style, onPageChange, onLoadProgress, ...others
}) => {
  const lastLoadedUrl = useRef<string>();

  const renderLoadingView = useCallback(() => {
    if (!renderLoading) return <View />;

    return (
      <View
        style={StyleSheet.absoluteFill}
      >
        {renderLoading()}
      </View>
    );
  }, [renderLoading]);

  const handleLoadProgress = useCallback((event: WebViewProgressEvent) => {
    onLoadProgress?.(event);

    const isPageChanged = lastLoadedUrl.current !== event.nativeEvent.url
      && event.nativeEvent.progress === 1;

    if (isPageChanged) {
      onPageChange?.({
        title: event.nativeEvent.title,
        url: event.nativeEvent.url,
      });

      lastLoadedUrl.current = event.nativeEvent.url;
    }
  }, [onLoadProgress, onPageChange]);

  const webViewStyles = useViewStyles(() => [
    styles.webView,
    style,
  ], [style]);

  return (
    <RNWebView
      startInLoadingState
      {...others}
      style={webViewStyles}
      renderLoading={renderLoadingView}
      onLoadProgress={handleLoadProgress}
    />
  );
};

export default WebView;
