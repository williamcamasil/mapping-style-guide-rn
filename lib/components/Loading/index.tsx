import React, { useEffect, useRef } from 'react';
import { StyleSheet, View, Animated } from 'react-native';

import { useViewStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';
import { getCircleBorderColor, getCircleBackgroundColor } from './utils';

export type LoadingPropsType = {
  theme: AppThemeType;
  circleSize?: number;
  color: ColorPaletesType;
};

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
  outCircle: {
    borderStyle: 'solid',
    overflow: 'hidden',
    marginHorizontal: 4,
  },
  circle: {
    width: '100%',
    height: '100%',
  },
});

const Loading: React.FC<LoadingPropsType> = ({ circleSize, color, theme }) => {
  const circleAnimation = useRef(new Animated.Value(0));

  const styleOutCircle = useViewStyles(
    () => [
      styles.outCircle,
      {
        borderColor: getCircleBorderColor(theme, color),
        width: circleSize,
        height: circleSize,
        borderRadius: circleSize && circleSize / 2,
        borderWidth: theme.borders.width.hairlene,
      },
    ],
    [circleSize, color, theme],
  );

  const styleCircle = useViewStyles(
    () => [
      styles.circle,
      {
        backgroundColor: getCircleBackgroundColor(theme, color),
        borderRadius: circleSize && circleSize / 2,
      },
    ],
    [theme, color, circleSize],
  );

  useEffect(() => {
    const animation = Animated.loop(
      Animated.timing(circleAnimation.current, {
        toValue: 50,
        duration: 1000,
        useNativeDriver: true,
      }),
    );
    animation.start();
    return () => {
      animation.stop();
    };
  }, []);

  const circleStyle1 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [1, 0.5, 0],
      }),
    }], [styleCircle]);

  const circleStyle2 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [0, 1, 0.5],
      }),
    }], [styleCircle]);

  const circleStyle3 = useViewStyles(() => [
    styleCircle,
    {
      opacity: circleAnimation.current.interpolate({
        inputRange: [0, 25, 50],
        outputRange: [0.5, 0, 1],
      }),
    }], [styleCircle]);

  return (
    <View style={styles.container}>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle1} />
      </View>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle2} />
      </View>
      <View style={styleOutCircle}>
        <Animated.View style={circleStyle3} />
      </View>
    </View>
  );
};

Loading.defaultProps = {
  circleSize: 8,
};

export default withTheme(Loading);
