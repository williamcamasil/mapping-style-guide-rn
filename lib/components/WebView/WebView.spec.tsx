import React from 'react';

import { render, fireEvent } from '@testing-library/react-native';

import WebView from '.';
import Container from '../Container';
import Loading from '../Loading';

describe('WebView snapshots', () => {
  it('default', () => {
    const result = render(<WebView source={{ uri: 'https://www.google.com.br/' }} />).toJSON();

    expect(result).toMatchSnapshot();
  });

  it('renderLoading', () => {
    const result = render(
      <WebView
        source={{ uri: 'https://www.google.com.br/' }}
        renderLoading={() => (
          <Container>
            <Loading color="neutralGray500" />
          </Container>
        )}
      />,
    ).toJSON();

    expect(result).toMatchSnapshot();
  });
});

describe('WebView events', () => {
  it('onPageChange', async () => {
    const handlePageChange = jest.fn();

    const result = render(
      <WebView
        testID="webview"
        source={{ uri: 'https://www.google.com.br/' }}
        onPageChange={handlePageChange}
      />,
    );

    fireEvent(result.getByTestId('webview'), 'onLoadProgress', {
      nativeEvent: {
        progress: 0.5,
        url: 'https://www.google.com.br/',
        title: 'google',
      },
    });

    fireEvent(result.getByTestId('webview'), 'onLoadProgress', {
      nativeEvent: {
        progress: 1,
        url: 'https://www.google.com.br/',
        title: 'google',
      },
    });

    fireEvent(result.getByTestId('webview'), 'onLoadProgress', {
      nativeEvent: {
        progress: 1,
        url: 'https://www.google.com.br/',
        title: 'google',
      },
    });

    expect(handlePageChange).toHaveBeenNthCalledWith(
      1,
      {
        url: 'https://www.google.com.br/',
        title: 'google',
      },
    );
  });

});
