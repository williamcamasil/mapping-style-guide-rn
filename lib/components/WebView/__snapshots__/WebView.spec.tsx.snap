// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebView snapshots default 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    cacheEnabled={true}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={true}
    onContentProcessDidTerminate={[Function]}
    onHttpError={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      {
        "uri": "https://www.google.com.br/",
      }
    }
    style={
      [
        {
          "flex": 1,
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#ffffff",
        },
        {
          "flex": 1,
        },
      ]
    }
    textInteractionEnabled={true}
    useSharedProcessPool={true}
  />
  <View />
</View>
`;

exports[`WebView snapshots renderLoading 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "overflow": "hidden",
      },
      undefined,
    ]
  }
>
  <RNCWebView
    cacheEnabled={true}
    injectedJavaScriptBeforeContentLoadedForMainFrameOnly={true}
    injectedJavaScriptForMainFrameOnly={true}
    javaScriptEnabled={true}
    messagingEnabled={true}
    onContentProcessDidTerminate={[Function]}
    onHttpError={[Function]}
    onLoadingError={[Function]}
    onLoadingFinish={[Function]}
    onLoadingProgress={[Function]}
    onLoadingStart={[Function]}
    onMessage={[Function]}
    onShouldStartLoadWithRequest={[Function]}
    source={
      {
        "uri": "https://www.google.com.br/",
      }
    }
    style={
      [
        {
          "flex": 1,
          "overflow": "hidden",
        },
        {
          "backgroundColor": "#ffffff",
        },
        {
          "flex": 1,
        },
      ]
    }
    textInteractionEnabled={true}
    useSharedProcessPool={true}
  />
  <View
    style={
      {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        {
          "alignItems": undefined,
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "flexDirection": undefined,
          "justifyContent": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "paddingVertical": undefined,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            {
              "borderColor": "hsla(223.60000000000002, 6.4%, 66.5%, 0.6)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "height": 8,
              "marginHorizontal": 4,
              "overflow": "hidden",
              "width": 8,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#A4A7AF",
                "borderRadius": 4,
                "height": "100%",
                "opacity": 1,
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "hsla(223.60000000000002, 6.4%, 66.5%, 0.6)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "height": 8,
              "marginHorizontal": 4,
              "overflow": "hidden",
              "width": 8,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#A4A7AF",
                "borderRadius": 4,
                "height": "100%",
                "opacity": 0,
                "width": "100%",
              }
            }
          />
        </View>
        <View
          style={
            {
              "borderColor": "hsla(223.60000000000002, 6.4%, 66.5%, 0.6)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "height": 8,
              "marginHorizontal": 4,
              "overflow": "hidden",
              "width": 8,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#A4A7AF",
                "borderRadius": 4,
                "height": "100%",
                "opacity": 0.5,
                "width": "100%",
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
</View>
`;
